pkg/dequeuer/async_handler.go

```
@@ -0,0 +1,206 @@
+	if err != nil {
+		return errors.Wrap(err, "failed to get payload")
+	}
+	defer h.deletePayload(requestID)
+
+	result, err := h.submitRequest(payload, requestID)
+	if err != nil {
+		return errors.Wrap(err, "failed to submit request to user container")
+	}
+
+	if err = h.uploadResult(requestID, result); err != nil {
+		return errors.Wrap(err, "failed to upload result to storage")
+	}
+
+	if err = h.updateStatus(requestID, status.AsyncStatusCompleted); err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to update status to %s", status.AsyncStatusCompleted))
+	}
+
+	h.log.Infow("workload processing complete", "id", requestID)
+
+	return nil
+}
+
+func (h *AsyncMessageHandler) handleFailure(requestID string) error {
+	err := h.updateStatus(requestID, status.AsyncStatusFailed)
+	if err != nil {
+		return errors.Wrap(err, fmt.Sprintf("failed to update status to %s", status.AsyncStatusFailed))
+	}
+	return nil
+}
+
+func (h *AsyncMessageHandler) updateStatus(requestID string, status status.AsyncStatus) error {
+	key := fmt.Sprintf("%s/%s/status/%s", h.storagePath, requestID, status)
+	return h.aws.UploadStringToS3("", h.config.Bucket, key)
+}
+
+func (h *AsyncMessageHandler) getPayload(requestID string) (*userPayload, error) {
+	key := fmt.Sprintf("%s/%s/payload", h.storagePath, requestID)
+	output, err := h.aws.S3().GetObject(
+		&s3.GetObjectInput{
+			Key:    aws.String(key),
+			Bucket: aws.String(h.config.Bucket),
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	contentType := "application/octet-stream"
+	if output.ContentType != nil {
+		contentType = *output.ContentType
+	}
+
+	return &userPayload{
+		Body:        output.Body,
+		ContentType: contentType,
+	}, nil
+}
+
+func (h *AsyncMessageHandler) deletePayload(requestID string) {
+	key := fmt.Sprintf("%s/%s/payload", h.storagePath, requestID)
+	err := h.aws.DeleteS3File(h.config.Bucket, key)
+	if err != nil {
+		h.log.Errorw("failed to delete user payload", "error", err) // TODO: telemetry
```

Telemetry todo?

---

we need to report the logged errors in general